File: ios/mmkv/SoloMMKVModule.m


#import <React/RCTBridgeModule.h>
#import "SoloMMKVModule.h"
#import <MMKV/MMKV.h>
#import <React/RCTLog.h>

@implementation SoloMMKVModule

RCT_EXPORT_MODULE(SoloMMKVModule)

RCT_EXPORT_BLOCKING_SYNCHRONOUS_METHOD(getString:(NSString *)key ) {
  MMKV *mmkv = [MMKV defaultMMKV];
  return [mmkv getStringForKey:key];
}

RCT_EXPORT_METHOD(setString:(NSString *)key value: (NSString *)value) {
  MMKV *mmkv = [MMKV defaultMMKV];
  [mmkv setString:value forKey:key];
}

RCT_EXPORT_METHOD(removeKey:(NSString *)key) {
  MMKV *mmkv = [MMKV defaultMMKV];
  [mmkv removeValueForKey:key];
}

RCT_EXPORT_METHOD(getStrings:
                  (NSString *)prefix
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
  MMKV *mmkv = [MMKV defaultMMKV];
  
  NSArray *allKeys = [mmkv allKeys];
  
  NSArray *values = [NSArray array];
  for (NSString *key in allKeys) {
    if ([key hasPrefix:(prefix)]) {
      values = [values arrayByAddingObject: [mmkv getStringForKey: key]];
    }
  }
  
  RCTLogInfo(@"Result %@", values);
  if (values) {
    resolve(values);
  } else {
    NSError *error = [NSError errorWithDomain:@"com.rachitmishra" code:204 userInfo:@"empty"];
    reject(@"empty", @"empty", error);
  }
}

RCT_EXPORT_METHOD(clearAll) {
  MMKV *mmkv = [MMKV defaultMMKV];
  [mmkv clearAll];
}



@end

File: android/app/src/main/java/com/rachitmishra/solo/mmkv/SoloMMKVModule.kt
package com.rachitmishra.solo.mmkv

import com.facebook.react.bridge.*
import com.tencent.mmkv.MMKV
import java.lang.NullPointerException
import java.lang.RuntimeException

class SoloMMKVModule(context: ReactApplicationContext) : ReactContextBaseJavaModule(context) {

    @ReactMethod(isBlockingSynchronousMethod = true)
    fun getString(key: String): String {
        val kv = MMKV.defaultMMKV()

        kv ?: return ""

        return kv.decodeString(key) ?: ""
    }

    @ReactMethod
    fun setString(key: String, value: String) {
        val kv = MMKV.defaultMMKV()
        kv?.encode(key, value)
    }

    @ReactMethod
    fun getStrings(prefix: String, promise: Promise) {
        val kv = MMKV.defaultMMKV()

        kv ?: return promise.reject(NullPointerException())

        val allKeys = kv.allKeys()

        allKeys ?: return promise.reject(NullPointerException())

        val allValues = allKeys.filter { key -> key.startsWith(prefix) }.map { key -> kv.decodeString(key) }

        if (allValues.isEmpty()) {
            return promise.reject(RuntimeException("empty"))
        }

        promise.resolve(allValues.toWritableArray())
    }

    private fun List<String?>.toWritableArray(): WritableArray {
        val result = Arguments.createArray()
        for (value in this) {
            result.pushString(value)
        }
        return result
    }

    @ReactMethod
    fun removeKey(key: String) {
        val kv = MMKV.defaultMMKV()
        kv?.removeValueForKey(key)
    }

    @ReactMethod
    fun clearAll() {
        val kv = MMKV.defaultMMKV()
        kv?.clearAll()
    }

    override fun getName(): String {
        return "SoloMMKVModule"
    }
}
